cmake/package_tooling.cmake

```diff
@@ -0,0 +1,128 @@
+#   $ cmake -DBOARD=<board> -B build -S samples/hello_world \
+#           -DMODULES=dts -P <ZEPHYR_BASE>/cmake/package_tooling.cmake
+#
+# It is also possible to pass additional build settings.
+# If you invoke CMake for 'hello_world' as:
+#
+#   $ cmake -DBOARD=<board> -B build -S samples/hello_world -DOVERLAY_CONFIG=foo.overlay
+#
+# you just add the same argument to tooling like:
+#   $ cmake -DBOARD=<board> -B build -S samples/hello_world -DOVERLAY_CONFIG=foo.overlay \
+#           -DMODULES=dts -P <ZEPHYR_BASE>/cmake/package_tooling.cmake
+#
+# Note: the samples CMakeLists.txt file is not processed by package tooling, so
+#       any 'set(<var> <value>)' specified before 'find_package(Zephyr)' must be
+#       manually applied, for example if the CMakeLists.txt contains:
+#          set(OVERLAY_CONFIG foo.overlay)
+#          find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+#       the 'foo.overlay' must be specified using '-DOVERLAY_CONFIG=foo.overlay'
+
+cmake_minimum_required(VERSION 3.20)
+
+# add_custom_target and set_target_properties are not supported in script mode.
+# However, several Zephyr CMake modules create custom target for user convenience
+# like menuconfig, boards, shields, etc.
+# As we are not generating a build system with this tool, only running part of
+# the modules, then we simply override those functions to allow running those
+# modules.
+function(add_custom_target)
+  # This silence the error: 'add_custom_target command is not scriptable'
+endfunction()
+
+function(set_target_properties)
+  # This silence the error: 'set_target_properties command is not scriptable'
+endfunction()
+
+# Find last `-B` and `-S` instances.
+foreach(i RANGE ${CMAKE_ARGC})
+  if(CMAKE_ARGV${i} MATCHES "^-B(.*)")
+    set(argB ${CMAKE_MATCH_1})
+    set(argB_index ${i})
+  elseif()
+  elseif(CMAKE_ARGV${i} MATCHES "^-S(.*)")
+    set(argS_index ${i})
+    set(argS ${CMAKE_MATCH_1})
+  endif()
+endforeach()
+
+if(DEFINED argB_index)
+  if(DEFINED argB)
+    set(CMAKE_BINARY_DIR ${argB})
+  else()
+    # value of -B follows in next index
+    math(EXPR argB_value_index "${argB_index} + 1")
+    set(CMAKE_BINARY_DIR ${CMAKE_ARGV${argB_value_index}})
+  endif()
+endif()
+
+if(DEFINED argS_index)
+  if(DEFINED argS)
+    set(APPLICATION_SOURCE_DIR ${argS})
+  else()
+    # value of -S follows in next index
+    math(EXPR argS_value_index "${argS_index} + 1")
+    set(APPLICATION_SOURCE_DIR ${CMAKE_ARGV${argS_value_index}})
+  endif()
+endif()
+
+# List of base modules needed in order to ensure same behavior as the default
+# Zephyr build system during configure time.
+set(BASE_MODULES python west root zephyr_module boards shields arch configuration_files)
+
+# dts is not directly a dependency of Kconfig, but in Zephyr build system we
+# import devicetree properties into the Kconfig tree and therefore we must ensure
+# dts module is loaded in Kconfig mode.
+if(kconfig IN_LIST MODULES)
+  list(APPEND BASE_MODULES dts)
+endif()
```

Comment:
Why doesn't dts include Kconfig?

Reply:
Because `dts` is run before `kconfig`, and hence, to run only the `dts` for Zephyr you don't need to run `kconfig`.

But if you want to run `kconfig` as is done in a regular Zephyr build, then you must run `dts` in order to have the devicetree knowledge available inside `kconfig` tree.

Reply:
Sorry, I wrote the dependency the wrong way. Can you answer again for the correct dependency direction?

Reply:
because the `kconfig` module itself doesn't depend on `dts`, but the way Zephyr Kconfig tree is written means that when Kconfig module is run in a Zephyr build, then `dts` is required.

This small difference becomes clearer when also looking at [SYSBuild](https://github.com/zephyrproject-rtos/zephyr/pull/40555), where SYSbuild re-uses Kconfig for it's menu, but doesn't re-use `dts`.
https://github.com/zephyrproject-rtos/zephyr/blob/e31348329a5e04f8c9293ac50bdb9918d2a238fe/sysbuild/CMakeLists.txt#L15-L16

This shows that Kconfig itself can be used without `dts`, but when wanting to run kconfig for a Zephyr build, then dts is required.
This is why this knowledge exists here, and not inside the kconfig CMake module.
Just like dts is loaded before kconfig here:
https://github.com/zephyrproject-rtos/zephyr/blob/184940230028729634828431073444cd97733874/cmake/modules/zephyr_default.cmake#L85-L86

Reply:
It would be unfortunate if we were overly generic for a future feature, and then that future feature doesn't get merged.

I had to revert lots of changes in master for the multi-image build because they were overly generic and then the
multi-image build was denied.

Can we do include(dts) now, and then, when we need to support non-DT invocation of Kconfig we make Kconfig agnostic to DT?

Reply:
> It would be unfortunate if we were overly generic for a future feature, and then that future feature doesn't get merged.

I know, that's why #40555 and the strategy behind it has already been presented both at the dev-review and the TSC, with very positive feedback.

And exactly due to that feedback and some existing additional use-cases, work like #41301 was started.

Reply:
Today, kconfig always uses DT, so it should include it as a dependency.

The future users that intend to decouple this dependency can do so in that PR.

Reply:
> Today, kconfig always uses DT, so it should include it as a dependency.

No, cause the Kconfig CMake module itself doesn't depend on dts.
The Kconfig tree does.

Just like the Kconfig CMake module doesn't depends on Zephyr modules, but our Kconfig tree uses it.

This is why there we have this:
https://github.com/zephyrproject-rtos/zephyr/blob/3df554fcb2c15c09278477bb4e04888d44683b19/cmake/modules/zephyr_default.cmake#L85-L86

That file is the order in which the Zephyr build itself expects things to be executed, but that is not the same as the Zephyr CMake modules themselves has such dependency.

We shouldn't artificial dependencies between modules, just because that's how Zephyr build system expects to execute them.

Reply:
A suggestion for getting rid of this dependency information has been proposed here[0], making the question of if kconfig should include dts moot.

[0] https://github.com/zephyrproject-rtos/zephyr/pull/41302/files#r797716740